Implement block tree for the domain v2

NingLin-P opened the pull request:
close #1610 

This PR implements block tree for the domain v2, which tracks the execution receipt for all domains.

This PR also has revised the `generate_genesis_state_root` host function to integrate the domain registry with block tree, but the genesis state root is not unique for every domain instance as the spec required. Because the genesis block builder takes runtime genesis config as an input and which is the same for all domain instances (the default value), adding digest to the genesis block header won't help as it won't affect the state root, we need to handle this issue in https://github.com/subspace/subspace/issues/1631.


### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

NingLin-P committed:
Add block tree specific runtime config

This commit also remove the unused ReceiptsPruningDepth

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Adjust the generate_genesis_state_root host function to remove the unneed runtime_genesis_config parameter

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Track successfully submitted bundles for all domains as now pallet-domains serve for all domains

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Add domain primitives for the domain v2 architecture

The v1 primitives can be removed and replaced by them  after the domain client
retired all of the v1 usage.

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Impl block tree for the domain v2

This commit includes the block tree types, storage items, the submit_bundle_v2 call and
pre_dispatch/validate_unsigned verification.

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Generate and import the genesis receipt to the block tree upon domain instantiation

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Revise and add test utilities for upcomming block tree test

Signed-off-by: linning <linningde25@gmail.com>

liuchengxu commented on crates/sp-domains/src/lib.rs:
```diff
@@ -475,6 +475,219 @@ where
     }
 }
 
+// Domain primitives for the v2 architecture
```
I suggest creating a new file `sp-domains/src/v2.rs` to reduce the potential conflicts as all of us can possibly change this file at the same time.

liuchengxu commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -116,6 +116,16 @@ mod pallet {
         /// Delay before a domain runtime is upgraded.
         type DomainRuntimeUpgradeDelay: Get<Self::BlockNumber>;
 
+        /// The block tree pruning depth, its value should >= `BlockHashCount` because we
```
This is a critical assumption, we can ensure that while building the runtime using `const_assert!(BlockTreePruningDepth::get() >= BlockHashCount::get())`.

NingLin-P replied:
There is a mistake, the assumption should be `BlockTreePruningDepth::get() <= BlockHashCount::get()`, and I found that this assumption is not enough to guarantee consensus block hash exists while verifying receipt because the domain block is not mapping to consensus block one by one, a domain can produce domain block 1 on consensus block 1 and produce domain block 2 at consensus block 1000...

Thus we may still need to manually store the consensus hash, but by utilizing `SuccessfulBundles` we can detect if there will be a domain block driving from the consensus block and only need to store the consensus block hash in this case. In the above case we only need to store consensus block hash of block 1 and block 1000 instead of all the hash from block 1 to block 1000 like v1 does.

liuchengxu replied:
That does not alleviate the problem honestly, it's expected that there are bundles in most consensus blocks for a domain, we run into the same rabbit hole as before if tracking the consensus block hash in the state, which was an open problem in v1 if you remember. We may have to introduce a host function to check it on the client side.

NingLin-P replied:
Yes, the problem still exists but is changed in some way under the v2 design.

Suppose we store hash for the domain if there is bundle in the consensus block, then for every consensus block:
1. if it contains no bundle then no block hash need to store
2. if all of its bundles are either confirming the current head or create fork in the block tree then it brings one hash to store
3. if it contains one bundle that extends the current head receipt then it brings one hash to store but also eliminates one
4. if it contains bundles that extend the block tree tip more than one (i.e. bundle1 extends head receipt, bundle2 extends bundle1's receipt, etc) then it eliminates more hash than it brings.

The 1st and 3rd cases are the most common while 2nd case is possible when there are network delays and malicious operators, 4th case is impossible currently unless we implement it :P 

Whenever the 2nd case brings an extra hash to store it also brings one gap between the block tree and the domain chain, because as the consensus block contains bundles it will bring one domain block to the domain chain but it doesn't extend the block tree at all. So the head receipt of the domain in the consensus chain can be block 10 while the best block of the domain chain already grows to block 100.

Similar to the hash storage the gap is also unbounded, which means unbounded confirmation of the domain chain, it should be something we need to fix and if it is fixed then the unbounded hash storage is also fixed.

liuchengxu replied:
The gap will be filled up eventually given the assumption that there is at least one honest operator IMO. The consensus chain needs to maintain a number of M * N consensus hash mappings (M = affected domains, N = produced domain blocks in this period) in this period of time, using a host function will free from maintaining these redundant mappings in the runtime.

NingLin-P replied:
> The gap will be filled up eventually given the assumption that there is at least one honest operator

It won't because ER is submitted with a bundle and there is only one ER within a bundle. In the best case, the ER extends the head receipt to reduce the gap by one but the bundle will result in a domain block thus increase the gap by one, so the gap remains.

My point is that the hash is used to verify upcoming ER in the first place, thus the hash is mapped to the domain block one by one like the ER, if the is no gap between the head ER and the best domain block then the number of hash is also bounded by `BlockTreePruningDepth`.

Assume `BlockTreePruningDepth` is 10, the head ER is block #50 and the best domain block is #N, the hashes we need to maintain are `block #40, ..., #N` if there is no gap `N == 50` then there are only 10 hashes need to maintain.

liuchengxu replied:
Yeah, I see, now that there is only one receipt per bundle, the gap can't be fixed once formed, this is another problem though. My point here is we don't have to track any consensus hashes in the state, they are redundant, and we can retrieve them from the client using a host function. We should be cautious of using the host function, but it's more straight here IMO.

NingLin-P replied:
When using the host function we will have to handle the chain fork on the client side which is also tricky. If we can fix the gap problem then the hashes storage is bounded by `BlockTreePruningDepth` per domain but it is indeed redundant.

liuchengxu replied:
Not necessarily, the runtime is tied to a specific block hash, so you don't have to deal with the forks.

NingLin-P replied:
But the client is not tied to any block or fork. I mean the only input to the host function is a block number, the client has to deal with forks and return a correct hash corresponding to the given block number.

liuchengxu replied:
You can extend the input using some block hash stored in the System pallet to help identify the exact branch and find the correct hash corresponding to the number.

No need to have a perfect solution in this PR, we can add a TODO, leave it as is, and revisit it later as the issue of the gap needs to be handled anyway.

NingLin-P replied:
> You can extend the input using some block hash stored in the System pallet to help identify the exact branch and find the correct hash corresponding to the number.

That is what I mean by tricky :P I will leave a TODO for now and open a new issue about the gap problem then we continue the discussion there.

liuchengxu commented on crates/pallet-domains/src/block_tree.rs:
```diff
@@ -0,0 +1,201 @@
+                let domain_block = maybe_domain_block.as_mut().expect(
+                    "The domain block of `CurrentHead` receipt is checked to be exist in `execution_receipt_type`; qed"
+                );
+                domain_block.operator_ids.push(submitter);
```
The duplicated operator_id is possible if pushed unconditionally.

vedhavyas replied:
Instead there the `operator_id` should be added only when the bundle they submitted is valid and unseen yet

NingLin-P replied:
Right, there are some comments on the `operator_ids` field that mentioned this, I think it is expected because an operator can submit multiple bundles within one consensus block and all of the bundle contains the same receipt that confirms the current head receipt. And since `operator_ids` is used to reward/slash operator, the reward/slash should account for every receipt the operator submitted.

liuchengxu replied:
It's possible to see one operator that submits multiple bundles in one consensus block, but it's meaningless to track duplicated operator_ids if the slashing is to [slash the entire pool]( https://www.notion.so/subspacelabs/Domains-v2-Specification-WIP-3fb0ec6e4d204c4881a7df50ef58da8f?pvs=4#962545ce4377412db4a6d114d4d73e47).

NingLin-P replied:
The operator pool is mapped to the operator one by one, I'm not sure if slashing the operator multiple times is desired in case of multiple bundles are submitted in one consensus block by the operator but rewarding the operator multiple times is needed in this case, thus we need to keep track of duplicated operator_id.

liuchengxu replied:
Doing the slashing multiple times is useless if the entire pool has been slashed the first time, but it makes sense for the rewarding purpose.

liuchengxu replied:
Resolving, but nice to add a bit more context to explain why duplicated operator_ids are expected.

liuchengxu commented on crates/sp-domains/src/lib.rs:
```diff
@@ -475,6 +475,219 @@ where
+    #[derive(Debug, Decode, Encode, TypeInfo, PartialEq, Eq, Clone)]
+    pub struct ExecutionReceipt<Number, Hash, DomainNumber, DomainHash, Balance> {
+        // The index of the current domain block that forms the basis of this ER.
+        pub domain_block_height: DomainNumber,
```
The spec has been updated to `domain_block_number` and `consensus_block_number` for consistency favor(the block_height is used less often than block_number).

liuchengxu commented on crates/sp-domains/src/lib.rs:
```diff
@@ -475,6 +475,219 @@ where
+        // The index of the current domain block that forms the basis of this ER.
+        pub domain_block_height: DomainNumber,
+        // A pointer to the hash of the ER for the last domain block.
+        pub parent_domain_block_receipt: H256,
```
The spec is updated to `parent_domain_block_receipt_hash`, which is more descriptive and consistent with some well-known concepts (`parent_block_hash`) IMHO.

liuchengxu reviewed (commented):
Some early comments, will take another look later.

Regarding the genesis extension, I have created an issue upstream https://github.com/paritytech/substrate/issues/14541. As Ved said, the quick workaround at present is to defer the genesis receipt generation to sometime later, we can always revisit once it's fixed upstream, and we can contribute to it for sure.

vedhavyas commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -574,8 +576,16 @@ mod pallet {
+        SuccessfulBundles::<T>::get(domain_id)
+    }
+
+    pub fn successful_bundles_of_all_domains() -> Vec<H256> {
```
Is this added to fetch all bundles at once instead of for specific domain? With storage map, do we have any performance benefit with this approach ?

NingLin-P replied:
Not for performance benefit but for use case in the bundle validator, which needs to keep track of all the successfully submitted bundles of the last K consensus block.

vedhavyas replied:
This is giving all the domain's budles. Why would every domain client would need to store these ?

NingLin-P replied:
The bundle validator exists in the consensus node not domain node.

liuchengxu replied:
The bundle validator was initially introduced to [prevent the duplicated bundles from being included in the consensus chain multiple times](https://github.com/subspace/subspace/issues/1056), specifically, bundle X is included in the consensus chain block 100, but the extrinsic `submit_bundle` in which bundle X was wrapped is still alive in the consensus network and it's possible to be included again in the consensus chain block 101 in the last version of block tree.

With the new block tree design, the duplicated bundle will be rejected as stale automatically as the receipt in it is required to either extend or confirm the block tree tip, the bundle validator is now unnecessary in this regard, we can consider even removing the entire bundle validator if no other concern.

NingLin-P replied:
> the duplicated bundle will be rejected as stale automatically as the receipt in it is required to either extend or confirm the block tree tip

I have also considered this but it may not be true because a duplicated bundle may still be considered valid if it confirms the tip, thus attackers can make use of that to keep submitting the bundle that contains the current head receipt (i.e when the head receipt change it also change to submit another bundle).

liuchengxu replied:
>thus attackers can make use of that to keep submitting the bundle that contains the current head receipt (i.e when the head receipt change it also change to submit another bundle).

I don't get this. A duplicated bundle that confirms the head receipt is never supposed to be included by an honest farmer because the duplicated bundle will be rejected as a duplicated transaction by the transaction pool, you can only send the bundles that are different, not duplicated.

NingLin-P replied:
Do you mean this check?https://github.com/paritytech/substrate/blob/e97e17445c8d354412de080c7ed3570dc0549799/client/transaction-pool/src/graph/validated_pool.rs#L145-L157

It seems the check is not guaranteed, `is_imported` only checks if tx is included in the ready queue after tx is included in the block the queue is cleared thus is not helpful in this case, and the `is_banned` has `ban_time` (default 30min, configurable) and the number of banned extrinsic limit (hardcoded 2048).

liuchengxu replied:
The check in the transaction pool ensures no duplicated bundles before they are included in the block, the runtime check ensures no duplicated bundles from the previous block can be included again. Once a bundle X is included in the block, and then you submit a bundle duplicate to X, which was created against a previous consensus block, it will be definitely rejected as stale in the runtime, no?

NingLin-P replied:
> the runtime check ensures no duplicated bundles from the previous block can be included again

I'm not sure if we have this check now, do you mean the check in `pre_dispatch/validate_unsigned` or the stale receipt check?

liuchengxu replied:
Hmm, sorry that I had a misunderstanding in the flow. I thought the bundles with a stale receipt will be rejected incorrectly, but they are not in order to calculate the bundle rate, so it does not work as I expected. We still need the bundle validator.

vedhavyas commented on crates/subspace-runtime/src/lib.rs:
```diff
@@ -792,13 +792,14 @@ impl_runtime_apis! {
 
         fn successful_bundle_hashes() -> Vec<H256> {
-            Domains::successful_bundles()
+            Domains::successful_bundles_of_all_domains()
```
I dont think this is correct. This func is called under the assumption that these belong to a single domain. So should be scoped to specific domain.

NingLin-P replied:
Not true, this func is called by bundle validator and is expected bundle of all domains, the use case you mentioned use `Domains::successful_bundles(domain_id)`.

vedhavyas replied:
Hmmm if the consensus node wants to fetch these bundles, which is the only case part that needs to see all domain bundles that I could think of, the result of this function is not very useful IMO.

NingLin-P replied:
> Hmmm if the consensus node wants to fetch these bundles, which is the only case part that needs to see all domain bundles that I could think of

This is the exact use case, we use the bundle validator in the consensus node's tx pool to detect duplicated bundles otherwise the attacker can use them to occupy the block space without cost.

vedhavyas commented on crates/pallet-domains/src/block_tree.rs:
```diff
@@ -0,0 +1,201 @@
+pub(crate) enum ReceiptType {
+    // New head receipt that extend the longest branch
+    NewHead,
+    // Receipt that comfirm the current head receipt
```
`comfirm` -> `confirms`

vedhavyas commented on crates/pallet-domains/src/block_tree.rs:
```diff
@@ -0,0 +1,201 @@
+    NewHead,
+    // Receipt that comfirm the current head receipt
+    CurrentHead,
+    // Receipt that create a new branch of the block tree
```
`creates`

vedhavyas commented on crates/pallet-domains/src/block_tree.rs:
```diff
@@ -0,0 +1,201 @@
+    CurrentHead,
+    // Receipt that create a new branch of the block tree
+    NewBranch,
+    // Receipt that newer than the head receipt but not extend the head receipt
```
`Receipt that is newer than the head receipt but does not extend the head receipt`

vedhavyas commented on crates/pallet-domains/src/block_tree.rs:
```diff
@@ -0,0 +1,201 @@
+        consensus_block_hash,
+        domain_block_height,
+        block_extrinsics_roots,
+        parent_domain_block_receipt,
```
`parent_domain_block_receipt_hash` is much more readable here

vedhavyas commented on crates/pallet-domains/src/block_tree.rs:
```diff
@@ -0,0 +1,201 @@
+    }
+
+    match execution_receipt_type::<T>(domain_id, execution_receipt) {
+        ReceiptType::InFuture => unreachable!("In future receipt should result in `UnknownParentBlockReceipt` error as it parent receipt is missing"),
```
IMO, I would rather return an error here and maybe log it if we are not expecting the error at this instant

vedhavyas commented on crates/pallet-domains/src/block_tree.rs:
```diff
@@ -0,0 +1,201 @@
+                    for block in BlockTree::<T>::take(domain_id, to_prune) {
+                        DomainBlocks::<T>::remove(block);
+                    }
+                    // Remove the block's `ExecutionInbox` as the block is pruned and not need
```
`Remove the block's `ExecutionInbox` as the block is pruned and do not need`

vedhavyas commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -368,6 +425,62 @@ mod pallet {
+                let next_number = HeadDomainNumber::<T>::get(domain_id)
+                    .checked_add(&One::one())
+                    .ok_or::<Error<T>>(BlockTreeError::MaxHeadDomainNumber.into())?;
+                HeadDomainNumber::<T>::set(domain_id, next_number);
```
I'm not sure about this. This should IMO be updated once we see the ER for these bundle. If the ER never arrives for any reason, then Consensus chain says its n+1 without any ER

NingLin-P replied:
> If the ER never arrives for any reason, then Consensus chain says its n+1 without any ER

It surely can, because a domain block must be driven from a consensus block that contains the domain's bundle. This is what the consensus chain can know for sure.

vedhavyas replied:
You are combining the state of Bundle submission and ER submission. If I think about primitives of the block chain itself, a head is only updated the execution of block at head is done. This is not the case here.

Clients may also fail at this point thinking Head is executed and for any proof verification and ends up failing instead. That is not a correct design IMO

NingLin-P replied:
> You are combining the state of Bundle submission and ER submission

Submission of bundle and ER is always combined in both v1 and v2.

> If I think about primitives of the block chain itself, a head is only updated the execution of block at head is done.

The `HeadDomainNumber` is stored in the consensus chain, suppose there is a consensus block that contains extrinsic that updates a domain 's`HeadDomainNumber` from 1 to 2. Operators who have processed this consensus block (drive & import a domain block) its best block number update from 1 to 2, which match your above statement.

 > Clients may also fail at this point thinking Head is executed and for any proof verification and ends up failing instead. That is not a correct design IMO

Confused here, essentially `HeadDomainNumber` is just a storage item in the consensus chain, it won't change something like `client.info().best_number`

vedhavyas replied:
> The HeadDomainNumber is stored in the consensus chain, suppose there is a consensus block that contains extrinsic that updates a domain 'sHeadDomainNumber from 1 to 2. Operators who have processed this consensus block (drive & import a domain block) its best block number update from 1 to 2, which match your above statement.

How do some one definitely prove that unless there is an execution receipt to corroborate that statement ?

NingLin-P replied:
"Drive a domain block if the consensus block contains bundles" is not unknown for the consensus node and is not something we need to prove.

vedhavyas replied:
I'm not sure if you understand my reasoning. Per my understanding, just submitting domain bundle is not the correct step to consider a domain chain has progressed unless there is ER associated with it. Lets take the current `DomainBlock`, It needs to have ER and per our definition of `DomainBlockNumber`, if there is a new domain block, then there is new head backing. Right now, we just do it on first bundle, which seem odd TBH.

NingLin-P replied:
Okay, I see the controversy here.

> just submitting domain bundle is not the correct step to consider a domain chain has progressed unless there is ER associated with it.

In my understanding, the domain chain is driven by the consensus chain so whenever it sees a consensus block that contains bundles, it drives and imports a domain block into the domain chain, so the chain progress at this moment. 

Due to the fact of the decouple execution architecture, the ER of the domain block is submitted to the consensus chain and tracked in the block tree, after the challenge period is passed the ER (namely the execution result of the domain block) is confirmed.

The `HeadReceiptNumber` is what we use to drive the challenge period moving forward thus it is apply to the whole block tree but not to any specific fork in the tree.

NOTE that the domain block **happen before** the ER, and the fork in the block tree is about the discrepancy of the execution result and is not the same as the chain fork. The term of `DomainBlock` and tree could be the source of confusion.

vedhavyas commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -368,6 +425,62 @@ mod pallet {
+
+            // Put the `extrinsics_root` to the inbox of the current under building domain block
+            let head_domain_number = HeadDomainNumber::<T>::get(domain_id);
+            ExecutionInbox::<T>::append(domain_id, head_domain_number, extrinsics_root);
```
Does this storage have no notion of forks? What if bundle was extending an fork?

NingLin-P replied:
Bundle does not have the notion of fork like ER as it does not have the notion of parent bundle.

In the block tree there are forks because the ER (or the resulting state root after execution) is unknown by the consensus chain since the bundle is executed by the operator. But here the consensus chain can know exactly which bundle is accepted in which consensus block and for a consensus block that contains bundles a domain block must be produced by the honest operator.

vedhavyas replied:
Ahh make sense Thanks!

vedhavyas commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -539,6 +655,16 @@ mod pallet {
+                    ValidTransaction::with_tag_prefix("SubspaceSubmitBundle")
+                        .priority(TransactionPriority::MAX)
+                        .longevity(T::ConfirmationDepthK::get().try_into().unwrap_or_else(|_| {
+                            panic!("Block number always fits in TransactionLongevity; qed")
```
use `expect` instead

NingLin-P replied:
This is port from existing code, using `expect` will result in compile error due to `<<T as frame_system::Config>::BlockNumber as TryInto<u64>>::Error` doesn't implement `Debug`

vedhavyas commented on crates/sp-domains/src/lib.rs:
```diff
@@ -508,6 +721,9 @@ impl PassBy for RuntimeType {
 pub type RuntimeId = u32;
 
+/// Type representing operator ID
+pub type OperatorId = u64;
```
I think already exists in the `main`. I think you need to rebase mayabe ?

NingLin-P replied:
Yes, the code is written before this exist in `main`

vedhavyas commented on crates/pallet-domains/src/tests.rs:
```diff
@@ -20,6 +20,9 @@ type Balance = u64;
 const DOMAIN_ID: DomainId = DomainId::new(0);
 
+// Operator id used for testing
+const OPERATOR_ID: OperatorId = 0u64;
```
Global constants will the root of all confusions when debugging tests. Every test should be self-contained so its easier to update then when needed

NingLin-P replied:
It is not the case here IMO, there is only one test using it because its value doesn't matter and just serves as a placeholder, even if we define it inside the test, the value will still be the same thus doesn't bring any help on debugging.

vedhavyas commented on crates/pallet-domains/src/tests.rs:
```diff
@@ -143,63 +149,94 @@ pub(crate) fn new_test_ext() -> sp_io::TestExternalities {
+
+pub(crate) struct ReadRuntimeVersion(pub Vec<u8>);
+
+impl sp_core::traits::ReadRuntimeVersion for ReadRuntimeVersion {
```
I dont understand the reason to move it here. We have it in runtime tests because that is the only test module that was using it. Right now, I dont see this being used anywhere else

vedhavyas commented on crates/pallet-domains/src/block_tree.rs:
```diff
@@ -0,0 +1,504 @@
+
+        let mut ext = new_test_ext();
+        ext.register_extension(sp_core::traits::ReadRuntimeVersionExt::new(
+            ReadRuntimeVersion(version.encode()),
```
Okay, I see this is used here. I'm still not sure how block tree tests need that when you can create the required state already.

Either way, would have preferred previous code move in this commit instead

NingLin-P replied:
> I'm still not sure how block tree tests need that when you can create the required state already.

The test call `Pallet::<Test>::register_domain_runtime` directly.

vedhavyas commented on crates/pallet-domains/src/block_tree.rs:
```diff
@@ -0,0 +1,504 @@
+    }
+
+    #[test]
+    fn test_block_tree() {
```
can we make this test more readable? It tries to test all the things in one test and very dense to actually track what is tested and what is not tested

NingLin-P replied:
Okay, I can split it into multiple tests but it may bring quite some duplicate setup code in each test.

vedhavyas replied:
This is how I did withdraw so I can test very specific scenaios. https://github.com/subspace/subspace/pull/1646/files#diff-bc31bd26a15ac3e6de45a5242e5fc7883136b699fa411a472d0d3eb45c50f0f9R775

Your test cases' mileage may very but I got minimal duplication for withdraws

NingLin-P replied:
`test_block_tree` is split into 6 tests, should be more readable now.

vedhavyas reviewed (commented):
Makes sense overall. Will take another look again.

NingLin-P committed:
Add block tree tests

Signed-off-by: linning <linningde25@gmail.com>

liuchengxu commented on crates/subspace-runtime/src/lib.rs:
```diff
@@ -472,8 +472,14 @@ parameter_types! {
     pub const DomainInstantiationDeposit: Balance = 100 * SSC;
     pub const MaxDomainNameLength: u32 = 32;
+    pub const BlockTreePruningDepth: u32 = 256;
+    pub const MaxBlockTreeFork: u32 = 32;
```
This is not defined in the spec, and I'm honestly unsure whether we should have such a constraint because it's different from the block tree on the client side, every fork receipt is expected to have a corresponding fraud proof and then the bad operator will be slashed. In addition, the fraud proof itself only handles the bad receipts and does not submit the good receipts, is it possible that someone keeps sending the bad receipt with higher fees which prevents the good receipts from being included?

liuchengxu commented on crates/pallet-domains/src/block_tree.rs:
```diff
@@ -0,0 +1,213 @@
+        ReceiptType::NewHead
+        | ReceiptType::NewBranch
+        | ReceiptType::CurrentHead
+        | ReceiptType::Stale => Ok(()),
```
Why Stale is Ok? Isn't it supposed to be an error?

NingLin-P replied:
According to [the spec](https://www.notion.so/subspacelabs/Domains-v2-Specification-WIP-3fb0ec6e4d204c4881a7df50ef58da8f?pvs=4#5a6f0e4cb0554aa3a4671cce1f5f596c), the stale receipt should not be added to the block tree but still need to track them for purposes of measuring the bundle production rate, this is handled inside the `submit_bundle_v2` call L539.

liuchengxu replied:
Resolving this as it will be more readable once https://github.com/subspace/subspace/pull/1650/files#r1262048802 is resolved.

liuchengxu commented on crates/pallet-domains/src/block_tree.rs:
```diff
@@ -0,0 +1,213 @@
+        // The genesis receipt is generated and added to the block tree by the runtime upon domain
+        // instantiation, thus it is unchallengeable and must always be the same.
+        ReceiptType::NewBranch if domain_block_number.is_zero() => {
+            Err(Error::ChallengeGenesisReceipt)
```
This case looks weird, it can be omitted by checking whether the genesis receipt matches the one on the consensus chain beforehand, and reject it directly as BadGenesisReceipt if they mismatch.

liuchengxu commented on crates/pallet-domains/src/block_tree.rs:
```diff
@@ -0,0 +1,213 @@
+
+    match receipt_number.cmp(&head_receipt_number.saturating_add(One::one())) {
+        Ordering::Greater => ReceiptType::InFuture,
+        Ordering::Equal => ReceiptType::NewHead,
```
This does not look correct, let's say the head receipt is N, a new receipt with the number N+1 does not mean it's a new head, the parent receipt hash of N+1 must also match the receipt hash N. That being said, `HeadReceiptNumber` is not enough, we should track the hash of the head receipt as well `HeadReceipt { number, hash }`.

```rust
if receipt_number == head_receipt.number + 1 
    && execution_receipt.parent_domain_block_receipt_hash == head_receipt.hash {
    // New head.
}
```

Same to the `CurrentHead`, the receipts with the same domain block number don't mean they confirm each other, you have to check the receipt hash.

Furthermore, I don't see any concept of block tree leaves, for the case of `NewBranch`, the new receipt should either confirm or extend that branch. The longest branch does not mean it's the _correct_ branch, the block tree needs to track all the branches and wait for the fraud proof to resolve which branch is the correct one.

liuchengxu replied:
I didn't notice the existence of `parent_domain_receipt_hash` has been checked in Line 122~126, this function makes sense now with that check. Also, the block tree leaves are not necessary after syncing up, the new head receipt can be discovered by checking if there is a receipt on each height while pruning the longest branch.

liuchengxu commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -465,6 +517,62 @@ mod pallet {
+
+            // Put the `extrinsics_root` to the inbox of the current under building domain block
+            let head_domain_number = HeadDomainNumber::<T>::get(domain_id);
+            ExecutionInbox::<T>::append(domain_id, head_domain_number, extrinsics_root);
```
Why not use `consensus_block_number` here, i.e., `ExecutionInbox = (domain_id, consensus_block_number, vec![bundle_extrinsics_root])`? Every domain block number has a corresponding consensus_block_number and it's already in the receipt. And then `HeadDomainNumber` is needless.

NingLin-P replied:
Using `consensus_block_number` make sense to me, thanks!

NingLin-P replied:
I found using `consensus_block_number` as the key of `ExecutionInbox` will bring some complexity to its removal. 

Namely, when a new head receipt is added to block tree, the oldest height of the block tree will be removed but if there are multiple ERs exist in that height and each ER points to different consensus blocks then it is not easy to tell which consensus block's `ExecutionInbox` to remove as if we remove the wrong one it will make valid ER of that consensus block unverifiable.

After considering serval edge cases I have come out with a solution that may fix this issue, namely to remove the `ExecutionInbox` of the consensus block with the smallest block number and if it is already removed then move to the second small block and so on. I can explain how the solution came out but as you can see it is more complicated and not straightforward compare to the current approach.

liuchengxu replied:
>each ER points to different consensus blocks

Actually, I think that is exactly another argument to use `consensus_block_number`, they shouldn't point to different consensus blocks. Given it's possible that there are multiple consensus blocks that could derive the same domain block when some of them may contain no bundles, let's say $Block_{C1}$ => $Block_{D1}$, $Block_{C2}$ => $Block_{D1}$,  ideally we should see $Receipt( C1, D1)$, but if someone submits a receipt $Receipt(C2, D1)$, when using `domain_block_number`, it will still be considered as valid if all the other fields are valid. If we use `consensus_block_number`, $Receipt(C2, D1)$ will be rejected as the execution inbox is not found. In other words, there is only one canonical receipt for a specific domain block, but there will be multiple using `domain_block_number` in theory, which is not ideal.

NingLin-P replied:
> If we use consensus_block_number, $Receipt(C2, D1)$ will be rejected as the execution inbox is not found

If $Block_{C2}$ also contains bundles then using `consensus_block_number` will also considered $Receipt(C2, D1)$ as valid. I think there are cases that can't be defended by using either `domain_block_number` or `consensus_block_number` and will have to involve fraud proof.

My main concern about using `consensus_block_number` is the removal of execution inbox. When there is a new head receipt added to the block tree, the ERs with domain block number `head_receipt_number - BlockTreePruningDepth` will be removed, when using `domain_block_number` we can know exactly which execution inbox to remove but when using `consensus_block_number` it is a lot more complicated.

liuchengxu replied:
No, $Block_{C2}$ does not contain any bundles, that's why C1 and C2 can derive the same domain block D1.

As I mentioned, it's possible that there are multiple valid receipts for a specific domain block at the same time when using `domain_block_number`, I'm unsure this's expected.

NingLin-P replied:
What I saying was the case that both $Block_{C1}$ and $Block_{C2}$ contain bundles and we expected $Receipt(C1, D1)$ and $Receipt(C2, D2)$, but using `consensus_block_number` can still considered $Receipt(C2, D1)$ as valid so it is not a silver bullet.

vedhavyas replied:
IMO, we should use  key as `(domain_id, consensus_block, domain_block)` so we keep track of all the extrinsincs roots seperately for each domain fork that might extend head or extend a branch. When an ER is submitted, these extrinsinc roots can be fetch based on the consensus block number it carries. So there should be 1 unqiue set of extrinsinc_roots for each concensus block and each fork of domain.

Storage pruning can be achieved with what you suggested with cleaning up the smallest consensus block

liuchengxu replied:
Well, that makes sense, then I think we need a pair of `(consensus_block_number, domain_block_number)` to obtain a canonical receipt for a domain block. (Didn't see Ved's comment when writing this :P )

NingLin-P replied:
I'm fine to use `(domain_id, consensus_block_number, domain_block_number)` as the key since it does help in some cases.

> so we keep track of all the extrinsincs roots seperately for each domain fork that might extend head or extend a branch.

Just to clarify, there is no domain chain fork regarding extrinsincs roots or bundle due to the fact that there is no consensus chain fork in the consensus runtime. The branches in the block tree are about the discrepancy between ER of the same domain block but not any chain fork.

NingLin-P committed:
Apply review suggestions

Signed-off-by: linning <linningde25@gmail.com>

vedhavyas commented on crates/pallet-domains/src/benchmarking.rs:
```diff
@@ -82,9 +82,9 @@ mod benchmarks {
     #[benchmark]
     fn submit_system_domain_invalid_state_transition_proof() {
-        let receipts_pruning_depth = T::ReceiptsPruningDepth::get().saturated_into::<u32>();
+        let receipts_pruning_depth = T::BlockTreePruningDepth::get().saturated_into::<u32>();
```
```suggestion
        let block_tree_pruning_depth = T::BlockTreePruningDepth::get().saturated_into::<u32>();
```

vedhavyas commented on crates/pallet-domains/src/block_tree.rs:
```diff
@@ -0,0 +1,700 @@
+) -> Result<(), Error> {
+    let er_hash = execution_receipt.hash();
+    match receipt_type {
+        er_type @ ReceiptType::NewHead | er_type @ ReceiptType::NewBranch => {
```
This does not look like rust 😉 
Did not know this existed. Might never use it 🤣

vedhavyas commented on crates/pallet-domains/src/domain_registry.rs:
```diff
@@ -63,13 +68,13 @@ impl DomainConfig {
     pub created_at: Number,
     /// The hash of the genesis execution receipt for this domain.
-    pub genesis_receipt_hash: Hash,
+    pub genesis_receipt_hash: H256,
```
why a hard type instead ?

NingLin-P replied:
`genesis_receipt_hash` is returned from the host function, which can't be generic type.

vedhavyas commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -455,6 +515,62 @@ mod pallet {
+            let operator_id = opaque_bundle.operator_id();
+            let (bundle_author, receipt) = opaque_bundle.deconstruct();
+
+            let receipt_type = execution_receipt_type::<T>(domain_id, &receipt);
```
I think this was called multiple times during the same function call IIUC

NingLin-P replied:
This is the only one inside the `submit_bundle_v2` as we pass the `receipt_type` to `process_execution_receipt`, the other 2 calls are in `pre_dispatch` and `validate_unsigned`

liuchengxu commented on crates/pallet-domains/src/block_tree.rs:
```diff
@@ -0,0 +1,700 @@
+            });
+        }
+        // Other types of receipt is unexpected for this function
+        _ => return Err(Error::UnexpectedReceiptType),
```
Since this function only expects certain types of receipts, if you refactor the ReceiptType to be more fruitful, this branch can be omitted.

```rust
enum AcceptedReceiptType {
    NewHead,
    NewBranch,
    CurrentHead,
}

enum RejectedReceiptType {
    InFuture,
    Pruned,
}

enum ReceiptType {
    Accepted(AcceptedReceiptType),
    Rejected(RejectedReceiptType),
    Stale
}
```

This function takes `AccepetedReceiptType` as the argument, the semantics is clearer as well.

NingLin-P replied:
Make sense to me!

liuchengxu reviewed (commented):
Mostly make sense, nice work!

vedhavyas commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -329,10 +380,16 @@ mod pallet {
-        /// Invalid bundle.
-        Bundle(BundleError),
+        /// Can not find the block hash of given primary block number.
+        UnavailablePrimaryBlockHash,
```
`UnavailableConsensusBlockHash`

vedhavyas commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -455,6 +515,62 @@ mod pallet {
+            // consensus block, which also mean a domain block will be produced thus update `HeadDomainNumber`
+            // to this domain block's block number.
+            if SuccessfulBundles::<T>::get(domain_id).is_empty() {
+                let next_number = HeadDomainNumber::<T>::get(domain_id)
```
another parallel question, why are we extending the head all receipt types? I'm not sure I see this handled. What if the first bundle is actually creating a new branch?
We should only extend for type `NewHead`, no ?

NingLin-P replied:
https://github.com/subspace/subspace/pull/1650#discussion_r1262581662 may answer this, although bundle and ER are submitted together, they are building different things. ER is used to build the block tree and may create a new branch in the tree while the bundle is used to build the domain chain.

vedhavyas commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -256,14 +286,58 @@ mod pallet {
+    pub(super) type DomainBlocks<T: Config> = StorageMap<
+        _,
+        Identity,
+        H256,
```
Can we please name this type? Its not easy to deduct what is used here ? And there is one in `BlockTree` storage as well.

vedhavyas commented on crates/pallet-domains/src/block_tree.rs:
```diff
@@ -0,0 +1,700 @@
+    execution_receipt: &ExecutionReceiptOf<T>,
+) -> ReceiptType {
+    let receipt_number = execution_receipt.domain_block_number;
+    let head_receipt_number = HeadReceiptNumber::<T>::get(domain_id);
```
I dont understant this. IIUC `head_receipt_number` should be the head of the fork this receipt is part of. But head receipt number you are using may or may not be the same

NingLin-P replied:
The `HeadReceiptNumber` is applied to the whole block tree not any specific fork in the tree, refer https://github.com/subspace/subspace/pull/1650#discussion_r1262581662.

vedhavyas commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -256,14 +286,58 @@ mod pallet {
+    /// number will be produce. Used as a pointer in `ExecutionInbox` to identify the current under building
+    /// domain block, also used as a mapping of consensus block number to domain block number.
+    #[pallet::storage]
+    pub(super) type HeadDomainNumber<T: Config> =
```
I'm not able to find specific need for this storage item and `HeadReceiptNumber`. Can you help me understand how these are useful ?

NingLin-P replied:
`HeadDomainNumber` is used to index the `ExecutionInbox` as the comment explains, for `HeadReceiptNumber` refer https://github.com/subspace/subspace/pull/1650#discussion_r1262581662.

vedhavyas reviewed (commented):
Left a few questions on few storage items. I'm not sure why we need them and those seems to cause more confusion to me actually

NingLin-P committed:
Apply review suggestions

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Merge branch 'main' of github.com:subspace/subspace into block-tree-pr

NingLin-P committed:
Fix domain tests

The domain tests are broken becasue the genesis domain is delay for 1 block
to initialize, thus the domain block can not produce block for the slot 1,
which cause the tests stuck. This commit fix it by produce 1 consensus block
upon setup to initialize the genesis domain.

Signed-off-by: linning <linningde25@gmail.com>

liuchengxu commented on crates/pallet-domains/src/block_tree.rs:
```diff
@@ -0,0 +1,716 @@
+
+    match execution_receipt_type::<T>(domain_id, execution_receipt) {
+        ReceiptType::Rejected(RejectedReceiptType::InFuture) => {
+            log::error!(
```
```suggestion
            log::error!(
                target: "runtime::domains",
```

liuchengxu commented on crates/pallet-domains/src/block_tree.rs:
```diff
@@ -0,0 +1,716 @@
+) -> Result<(), Error> {
+    let er_hash = execution_receipt.hash();
+    match receipt_type {
+        er_type @ AcceptedReceiptType::NewHead | er_type @ AcceptedReceiptType::NewBranch => {
```
Having fewer branches looks more readable to me, non-blocker.

```rust
    let er_hash = execution_receipt.hash();
    let domain_block_number = execution_receipt.domain_block_number;
    let add_new_receipt_to_block_tree = || {
        // Construct and add a new domain block to the block tree
        let domain_block = DomainBlock {
            execution_receipt,
            operator_ids: sp_std::vec![submitter],
        };
        BlockTree::<T>::mutate(domain_id, domain_block_number, |er_hashes| {
            er_hashes.insert(er_hash);
        });
        DomainBlocks::<T>::insert(er_hash, domain_block);
    };
    match receipt_type {
        AcceptedReceiptType::NewBranch => add_new_receipt_to_block_tree(),
        AcceptedReceiptType::NewHead => {
            add_new_receipt_to_block_tree();

            // Update the head receipt number
            HeadReceiptNumber::<T>::insert(domain_id, domain_block_number);

            // Prune expired domain block
            if let Some(to_prune) =
                domain_block_number.checked_sub(&T::BlockTreePruningDepth::get())
            {
                for block in BlockTree::<T>::take(domain_id, to_prune) {
                    DomainBlocks::<T>::remove(block);
                }
                // Remove the block's `ExecutionInbox` as the block is pruned and does not need
                // to verify its receipt's `extrinsics_root` anymore
                let _ = ExecutionInbox::<T>::clear_prefix((domain_id, to_prune), u32::MAX, None);
            }
        }
        AcceptedReceiptType::CurrentHead => {
......
        }
    }
```

NingLin-P replied:
Don't want to introduce a stateful closure here thus it is replaced by a function.

liuchengxu commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -755,11 +853,71 @@ mod pallet {
+            let db_weight = T::DbWeight::get();
+            db_weight
+                .reads(results.loops as u64)
+                .saturating_add(db_weight.writes(1))
```
This can be simply ignored actually for now as the weights for the entire function are still incorrect even with this effort.

liuchengxu commented on crates/pallet-domains/src/block_tree.rs:
```diff
@@ -0,0 +1,716 @@
+            Err(Error::InFutureReceipt)
+        }
+        ReceiptType::Rejected(RejectedReceiptType::Pruned) => {
+            log::error!(
```
```suggestion
            log::error!(
                target: "runtime::domains",
```

liuchengxu reviewed (commented):
looks good to me to proceed, a few minor comments, none of them are blockers.

Regarding whether we should take all bundles that are possibly from different domain branches as the input for the next domain block, we can revisit once it's more clearly defined in the spec.

vedhavyas commented on crates/pallet-domains/src/block_tree.rs:
```diff
@@ -0,0 +1,716 @@
+    match receipt_number.cmp(&head_receipt_number.saturating_add(One::one())) {
+        Ordering::Greater => ReceiptType::Rejected(RejectedReceiptType::InFuture),
+        Ordering::Equal => ReceiptType::Accepted(AcceptedReceiptType::NewHead),
+        Ordering::Less => {
```
not a blocker but would be great to improve this arm. One too many `if..else if` IMO

NingLin-P replied:
`if .. else if` can be replaced by a bunch of `if .. { return Error::xx }` but logically it is the same.

vedhavyas replied:
🤷🏼‍♂️

vedhavyas reviewed (commented):
Though I'm still skeptical about the assumptions made specifically on bundles, I think we can revisit them once we have some concrete examples that might warrant for a refactor.

Please fix conflicts and we can get this in 👍🏼 
Thank you for being patient @NingLin-P

NingLin-P committed:
Apply review suggestions

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Merge branch 'main' of github.com:subspace/subspace into block-tree-pr

NingLin-P committed:
Merge branch 'main' of github.com:subspace/subspace into block-tree-pr
